### PR TITLE
Add shutdown handler for new launcher

### PR DESCRIPTION
**Pull request by rvl**
Relates to #1314.

# Overview

This adds a simpler way of ensuring clean shutdown of the wallet on windows (and linux), which doesn't require DaedalusIPC. The mechanism will be used by cardano-launcher.

- Adds the shutdown handler thread.
- Needed to rearrange startup functions around a little bit.
- Unit tests and an integration test.

# Comments

[Hydra jobset](https://hydra.iohk.io/jobset/Cardano/cardano-wallet-pr-1315)

### CODE REVIEW

File: default.nix [hunk 1]
```diff
@@ -39,6 +39,8 @@ let
       haskellBuildUtils = iohkLib.haskellBuildUtils.package;
     };
 
+    inherit (haskellPackages.cardano-wallet-byron.components.exes) cardano-wallet-byron;
```
Comment (rvl) [inline default.nix:34]: This is used by shell.nix in cardano-launcher.

**Comment by rvl** [comment]
bors try

File: lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/CLI/Server.hs [hunk 2]
```diff
@@ -237,6 +244,30 @@ spec = do
...
+                        , "--random-port"
+                        , "--genesis-block-hash", block0H
+                        ]
+                        (pure ())
+                        CreatePipe
+                        stream
+
+                withBackendProcessHandle nullTracer cmd $ \(Just hStdin) _ -> do
+                    threadDelay oneSecond
+                    hClose hStdin
```
Comment (paweljakubas) [inline lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/CLI/Server.hs:264]: **[question]**

it all comes down to this when shutting down on cmd, right? No way that we have some kind of shutdown without closing stdin handle?

**Review by paweljakubas — APPROVED** [review]
I was looking only at Haskell code only trying to understand logic, mainly in Startup and Launcher modules. Looks fine. Very fond of different cases covered in Cardano.StartupSpec. Looks like you nail it. Of course the ultimate verification is upon operation which to be seen.
btw. I see some complaining of weeder : 
```
warning: unknown setting 'sandbox-fallback'
```

File: lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/CLI/Server.hs (hunk 2 above)
Comment (rvl) [reply lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/CLI/Server.hs:264]: Yes good point - I added a check to make sure that the process exits with status 0.

**Comment by rvl** [comment]
I fixed weeder and rebased.
It works well on Linux with the cardano-launcher code.
On Windows it deadlocks because it basically does this `race (BS.hGet stdin 1000) (threadDelay 1000000)`, which also deadlocks on Windows. I haven't found a workaround yet.

**Review by KtorZ — APPROVED** [review]

**Pull request merged**